### PR TITLE
Fixing unreasonable memory usage when creating an InferenceData object from an emcee sampler

### PR DESCRIPTION
**Pull request by acpetit**
## Description

The goal of this change is to prevent large memory consumption during the creation of an InferenceData object from emcee. The change also improves efficiency.

Previously, the full chain was initialized once for each of the variables, leading to long initialization time and a spike in memory usage (my 2.5Gb chain with ~30 dimensions led to 60+Gb of allocated memory, crashing the python process).

## Checklist
- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md#pull-request-checklist) PR format
- [x] New features are properly documented (with an example if appropriate)?
- [x] Code style correct (follows pylint and black guidelines)
- [x] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/main/CHANGELOG.md#v0xx-unreleased)

----
:books: Documentation preview :books:: https://arviz--2215.org.readthedocs.build/en/2215/

### CODE REVIEW

File: arviz/data/io_emcee.py [hunk 1]
```diff
@@ -115,13 +115,15 @@ def __init__(
 
     def posterior_to_xarray(self):
         """Convert the posterior to an xarray dataset."""
+        # A chain object is created outside of the dictionary to avoid a spike in memory usage
```
Comment (OriolAbril) [inline arviz/data/io_emcee.py:118]:
```suggestion
```

File: arviz/data/io_emcee.py [hunk 2]
```diff
@@ -115,13 +115,15 @@ def __init__(
...
         # Use emcee3 syntax, else use emcee2
+        if hasattr(self.sampler, "get_chain"):
+            chain = self.sampler.get_chain().swapaxes(0, 1)
```
Comment (OriolAbril) [inline arviz/data/io_emcee.py:121]:
```suggestion
            samples_ary = self.sampler.get_chain().swapaxes(0, 1)
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 # Change Log
 
+## v0.x.x Unreleased
+
+### Maintenance and fixes
+- Fix memory usage and improve efficiency for `io_emcee.posterior_to_xarray`.
```
Comment (OriolAbril) [inline CHANGELOG.md:6]:
```suggestion
- Fix memory usage and improve efficiency in `from_emcee`.
```

**Review by OriolAbril — APPROVED** [review]
left two minor comments about using `from_emcee` to keep the changelog user friendly (hopefully) as users don't need to know about io_emcee, also rename `chain` to `samples_ary` as that array contains samples from all chains.

Approving already though, that was a very good catch, thanks!

**Comment by acpetit** [comment]
Thanks ! :)

**Pull request merged**